zebra-state/src/utxo.rs

```diff
@@ -0,0 +1,42 @@
+// needed to make clippy happy with derive(Arbitrary)
+#![allow(clippy::unit_arg)]
+
+use zebra_chain::{block, transparent};
+
+/// An unspent `transparent::Output`, with accompanying metadata.
+#[derive(Clone, Debug, PartialEq, Eq)]
+#[cfg_attr(test, derive(proptest_derive::Arbitrary))]
+pub struct Utxo {
+    /// The output itself.
+    pub output: transparent::Output,
+    /// The height at which the output was created.
+    pub height: block::Height,
+    /// Whether the output originated in a coinbase transaction.
+    pub from_coinbase: bool,
+}
+
+#[cfg(test)]
+pub fn new_outputs(block: &block::Block) -> std::collections::HashMap<transparent::OutPoint, Utxo> {
+    use std::collections::HashMap;
+
+    let height = block.coinbase_height().expect("block has coinbase height");
+
+    let mut new_outputs = HashMap::default();
+    for transaction in &block.transactions {
+        let hash = transaction.hash();
+        let from_coinbase = transaction.is_coinbase();
+        for (index, output) in transaction.outputs().iter().cloned().enumerate() {
+            let index = index as u32;
+            new_outputs.insert(
+                transparent::OutPoint { hash, index },
+                Utxo {
+                    output,
+                    height,
+                    from_coinbase,
+                },
```

Comment:
Do we also want to add the transaction index in the block to `Utxo`, so we can efficiently lookup the transaction in the block?

(Some blocks have thousands of transactions.)

Reply:
I didn't add the transaction index to try to keep the size of the UTXO data down; we never need to look up the transaction in the course of normal node operations, only potentially for RPC queries from external software (e.g., a wallet looking up the status of one of its UTXOs), so I think it's fine for something to be possible-but-slow and to build more detailed indexes if there was a need to do so.